src/compressed_chunk.c

```diff
@@ -33,6 +42,129 @@ void Compressed_FreeChunk(Chunk_t *chunk) {
+    size_t curNumSamples = curChunk->count - split;
+
+    // add samples in new chunks
+    size_t i = 0;
+    Sample sample;
+    ChunkIter_t *iter = Compressed_NewChunkIterator(curChunk, false);
+    CompressedChunk *newChunk1 = Compressed_NewChunk(curChunk->size / sizeof(Sample));
+    CompressedChunk *newChunk2 = Compressed_NewChunk(curChunk->size / sizeof(Sample));
+    for (; i < curNumSamples; ++i) {
+        Compressed_ChunkIteratorGetNext(iter, &sample);
+        if (Compressed_AddSample(newChunk1, &sample) != CR_OK) {
+            goto error;
+        }
+    }
+    for (; i < curChunk->count; ++i) {
+        Compressed_ChunkIteratorGetNext(iter, &sample);
+        if (Compressed_AddSample(newChunk2, &sample) != CR_OK) {
+            goto error;
+        }
+    }
+
+    trimChunk(newChunk1);
+    trimChunk(newChunk2);
+    swapChunks(curChunk, newChunk1);
+
+    Compressed_FreeChunkIterator(iter, false);
+    Compressed_FreeChunk(newChunk1);
+
+    return newChunk2;
+
+error:
+    Compressed_FreeChunkIterator(iter, false);
+    Compressed_FreeChunk(newChunk1);
+    Compressed_FreeChunk(newChunk2);
+    return NULL;
+}
+
+ChunkResult Compressed_UpsertSample(AddCtx *aCtx) {
+    ChunkResult rv = CR_OK;
+    ChunkResult res = CR_OK;
+    CompressedChunk *oldChunk = (CompressedChunk *)aCtx->inChunk;
+
+    short newSize = oldChunk->size / sizeof(Sample);
+    // extend size if approaching end
+    if (((oldChunk->idx) / BIT) + EXTRA_SPACE >= oldChunk->size) {
+        newSize += EXTRA_SPACE / sizeof(Sample);
+    };
+
+    CompressedChunk *newChunk = Compressed_NewChunk(newSize);
+    Compressed_Iterator *iter = Compressed_NewChunkIterator(oldChunk, false);
+    timestamp_t ts = aCtx->sample.timestamp;
+    int numSamples = oldChunk->count;
+
+    size_t i = 0;
+    Sample iterSample;
+    for (; i < numSamples; ++i) {
+        res = Compressed_ChunkIteratorGetNext(iter, &iterSample);
+        if (iterSample.timestamp >= ts) {
+            break;
+        }
+        res = Compressed_AddSample(newChunk, &iterSample);
+        CHECK_EXTEND_CHUNK(res, newChunk, iterSample);
+    }
+
+    if (ts == iterSample.timestamp) {
+        res = Compressed_ChunkIteratorGetNext(iter, &iterSample);
+        aCtx->sz = -1;                     // we skipped a sample
+    } else if (aCtx->type == UPSERT_DEL) { // No sample to delete
+        rv = CR_DEL_FAIL;
+        goto clean;
+    }
+
+    if (aCtx->type != UPSERT_DEL) {
+        ChunkResult resSample = Compressed_AddSample(newChunk, &aCtx->sample);
+        CHECK_EXTEND_CHUNK(resSample, newChunk, aCtx->sample);
+        aCtx->sz += 1;
+    }
+
+    if (i != numSamples) { // sample is not last
+        while (res == CR_OK) {
+            res = Compressed_AddSample(newChunk, &iterSample);
+            CHECK_EXTEND_CHUNK(res, newChunk, iterSample);
+            res = Compressed_ChunkIteratorGetNext(iter, &iterSample);
+        }
+    }
+
+    // trim data
+    if (!aCtx->latestChunk) {
```

Comment:
Since we recreate the chunk, for old chunks we want to release any excess memory back to the system.
If it is the latest chunk, we want to keep the space for upcoming TS.ADD commands so we keep the original chunk size.

Reply:
It might get repeating extra spaces and grow over time